File: 025.Reverse_Nodes_i_k-Group/Solution.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
    if (head == null)
        return null;
    ListNode point = head;
    //找到子链表的尾部
    int i = k;
    while(i - 1 >0){
        point = point.next;
        if (point == null) {
            return head;
        }
        i--;
    }
    ListNode temp = point.next;
     //将子链表断开
    point.next = null;

    //倒置子链表，并接受新的头结点
    ListNode new_head = reverse(head);

    //head 其实是倒置链表的尾部，然后我们将后边的倒置结果接过来就可以了
    //temp 是链表断开后的头指针，可以参考解法一的图示
    head.next = reverseKGroup(temp,k);
    return new_head;
}
public ListNode reverse(ListNode head) {
    ListNode current_head = null;
    while (head != null) {
        ListNode next = head.next;
        head.next = current_head;
        current_head = head;
        head = next;
    }
    return current_head;
}
}

File: 025.Reverse_Nodes_i_k-Group/Solution.py
class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        def reverseLinkedList(node):
            prev, curr = None, node
            while curr is not None:
                tmp = curr.next
                curr.next = prev
                prev = curr
                curr = tmp
            return prev, node
        
        def getKNodes(node, k):
            curr = node
            while k > 1 and curr is not None:
                k -= 1
                curr = curr.next
            if curr is None:
                return False, curr
            return True, curr
        
        if k < 2:
            return head
        nh = ListNode(None)
        nh.next = head
        curr = nh
        while curr is not None:
            part_head = curr.next
            can_take, part_tail = getKNodes(part_head, k)
            if not can_take:
                break
            next_head = part_tail.next
            part_tail.next = None
            rh, rt = reverseLinkedList(part_head)
            curr.next = rh
            curr = rt
            curr.next = next_head
        return nh.next
